### PR TITLE
helix: Support `[[grammar]]` in languages.toml

### LINKED ISSUE #2871: Helix: support configuring grammars
### Support configuring tree-sitter grammars in `languages.toml` for Helix.

Docs: https://docs.helix-editor.com/languages.html#tree-sitter-grammars

```toml
[[grammar]]
name = "rust"
source = { git = "https://github.com/tree-sitter/tree-sitter-rust", rev = "a250c4582510ff34767ec3b7dcdd3c24e8c8aa68" }
```

The `use-grammars` key should also be supported, which controls which grammars are fetched and built (if omitted all grammars are used).

```toml
use-grammars = { only = [ "rust", "c", "cpp" ] }
# or
use-grammars = { except = [ "yaml", "json" ] }
```

### PR DESCRIPTION
**Pull request by matthewess**
### Description

Add support for the `[[grammar]]` key of `languages.toml`: https://docs.helix-editor.com/languages.html#tree-sitter-grammar-configuration

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

edit: Should fix #2871

### CODE REVIEW

File: tests/modules/programs/helix/languages-expected-use-grammars.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[use-grammars]
+except = ["yaml", "json"]
+[[grammar]]
```
Comment (matthewess) [inline tests/modules/programs/helix/languages-expected-use-grammars.toml:3]: i dont love that using `concatTextFile` gives us this very slightly ugly result without an additional break between sections, unsure if there is a workaround

**Comment by matthewess** [comment]
cc @Philipp-M in case you have any thoughts

File: modules/programs/helix.nix [hunk 2]
```diff
@@ -136,10 +173,29 @@ in {
...
-          source =
-            tomlFormat.generate "helix-config" { language = cfg.languages; };
-        };
+        "helix/languages.toml" = mkIf (cfg.languages != [ ] || cfg.grammars
+          != [ ] || cfg.use-grammars != { }) {
+            source =
+              # NB: helix requires that `use-grammars` be the first key in languages.toml if present
+              # pkgs.formats.toml relies on piping generated json through
+              # [remarshal](https://github.com/remarshal-project/remarshal),
+              # so we have to concatenate in order to preserve this ordering
```
Comment (Philipp-M) [inline modules/programs/helix.nix:182]: Are you sure about that? AFAIK helix uses the `serde_derive` implementation for it which should respect the TOML standard, no weird TOML breaking custom deserializer (in the way that it relies on ordered tables/objects).

**Comment by Philipp-M** [comment]
Sorry almost forgot about it.

I'm not sure about the PR (see comment in code), I think it's better to just serialize the whole languages.toml.

I have implemented this some time back in https://github.com/Philipp-M/home-manager/commit/8bdd7cc552acb2f78391d82cccdfc0b306fbc623.
Actually I have planned to upstream this when a new helix release is out that has merged this: https://github.com/helix-editor/helix/pull/2507

But it doesn't solve the grammars issue see the discussion in #2871

File: modules/programs/helix.nix (hunk 2 above)
Comment (ncfavier) [reply modules/programs/helix.nix:182]: The docs say
> `# Note: this key must come **before** the [[language]] and [[grammar]] sections`
but indeed this just means that it should be *outside* of the `language` and `grammar` sections, so just regular TOML.

**Pull request closed without merging**